src/Symfony/Component/Routing/Loader/Psr4DirectoryLoader.php

```diff
@@ -12,35 +12,36 @@
 namespace Symfony\Component\Routing\Loader;
 
 use Symfony\Component\Config\FileLocatorInterface;
-use Symfony\Component\Config\Loader\FileLoader;
+use Symfony\Component\Config\Loader\Loader;
 use Symfony\Component\Routing\RouteCollection;
 
 /**
  * A loader that discovers controller classes in a directory that follows PSR-4.
  *
  * @author Alexander M. Turek <me@derrabus.de>
  */
-final class Psr4DirectoryLoader extends FileLoader
+final class Psr4DirectoryLoader extends Loader
 {
-    public function __construct(FileLocatorInterface $locator)
-    {
+    public function __construct(
+        private readonly FileLocatorInterface $locator,
+    ) {
         // PSR-4 directory loader has no env-aware logic, so we drop the $env constructor parameter.
-        parent::__construct($locator);
+        parent::__construct();
     }
 
     public function load(mixed $resource, string $type = null): ?RouteCollection
     {
-        $path = $this->locator->locate($resource);
+        $path = $this->locator->locate($resource['path']);
         if (!is_dir($path)) {
             return new RouteCollection();
         }
 
-        return $this->loadFromDirectory($path, trim(substr($type, 10), ' \\'));
+        return $this->loadFromDirectory($path, trim($resource['namespace'], '\\'));
     }
 
     public function supports(mixed $resource, string $type = null): bool
     {
-        return \is_string($resource) && null !== $type && str_starts_with($type, 'attribute@');
+        return 'attribute' === $type && \is_array($resource) && isset($resource['path'], $resource['namespace']);
```

Comment:
We could probably support the `annotations` type for free here, by allowing both (and forwarding the type in `loadFromDirectory`)